### PR TITLE
Warn if user is using <title> within the _document.js

### LINKED ISSUE #4596: Using <title> in _document.js's <Head> tag should show a warning to use _app.js instead
# Bug report

Based on #3527

## Describe the bug

Currently, users might add `<title>` in `_document.js` which will lead to unexpected results with `next/head` since `_document.js` is only rendered on the initial pre-render. `_app.js` is rendered on the server and client side, and is rendered as a normal React tree, whereas _document.js is rendered as static HTML.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:
1. Copy the default `_document.js` and add `<title>`:
```
// _document is only rendered on the server side and not on the client side
// Event handlers like onClick can't be added to this file

// ./pages/_document.js
import Document, { Head, Main, NextScript } from 'next/document'

export default class MyDocument extends Document {
  static async getInitialProps(ctx) {
    const initialProps = await Document.getInitialProps(ctx)
    return { ...initialProps }
  }

  render() {
    return (
      <html>
        <Head>
          <title>Hello World</title>
        </Head>
        <body className="custom_class">
          <Main />
          <NextScript />
        </body>
      </html>
    )
  }
}
```

## Expected behavior

Above example should show a warning that `<title>` should only be used in _app.js and other pages. This can be done using `err.sh`, create a link like: `https://err.sh/next.js/no-document-title` and create a file in the `errors` directory: `errors/no-document-title.md`

### PR DESCRIPTION
**Pull request by davscro**
resolves #4596 
@timneutkens I'm not sure is it right place to call this function in ```render``` method or IMHO it's better to place it within the ```constructor```, what do u think?

### CODE REVIEW

**Review by timneutkens — CHANGES_REQUESTED** [review]
Same feedback as before: https://github.com/zeit/next.js/pull/4519#issuecomment-398563504

**Comment by davscro** [comment]
but ```{(children || []).map((child) => {``` is not correct as children could be a single object not an array 😉

**Comment by timneutkens** [comment]
You can use https://reactjs.org/docs/react-api.html#reactchildrentoarray

**Comment by davscro** [comment]
Cool, I didn't know about React.Children utils, that's why I did custom implementation. Thanks for the hint.

**Comment by timneutkens** [comment]
I wonder if we can disable this in production / compile away the code

**Comment by timneutkens** [comment]
There is no need for that, just use `if(process.env.NODE_ENV === 'development')`

File: server/document.js [hunk 1]
```diff
@@ -113,7 +113,12 @@ export class Head extends Component {
       {this.getPreloadMainLinks()}
       {this.getCssLinks()}
       {styles || null}
-      {this.props.children}
+      {process.env.NODE_ENV === 'development' && (React.Children.toArray(this.props.children)).map((child) => {
```
Comment (lucleray) [inline server/document.js]: Am I missing something or `this.props.children` won't be rendered at all if `NODE_ENV` is not `production` ? 😅

Comment (timneutkens) [reply server/document.js]: Correct, this is wrong, and uglify won't remove the block if it's not development. Instead we have to do something like this:

```
let children
if (process.env.NODE_ENV === 'development') {
 children = // etc etc
} else {
 children = this.props.children
}
```

This way uglify will remove the specific block

Comment (lucleray) [reply server/document.js]: Why not just use `React.Children.forEach` to show the warning ? And keep `{this.props.children}` ?
```js
if (process.env.NODE_ENV === 'development') {
  React.Children.forEach(this.props.children, (child) => {
    if (child.type === 'title') {
      console.warn("...")
    }
  })
}
```

Comment (timneutkens) [reply server/document.js]: Doing 1 loop instead of 2 🤔

Comment (lucleray) [reply server/document.js]: But only in `development` 🙈

Comment (Jauny) [reply server/document.js]: Random question - is it not better practice also to check process.env.NODE_ENV !== 'production' instead? I thought that it was safer to assume that 'production' would be present in prod. Some people might want warnings to show in 'staging' also for example.

Comment (lucleray) [reply server/document.js]: Yes ! That's also how it was implemented 🙂

https://github.com/zeit/next.js/pull/5160#discussion_r219099014

**Pull request closed without merging**